Adding documentation for Topology Aware Hints (#27032)
kubernetes/website: content/en/docs/concepts/services-networking/topology-aware-hints.md

---
reviewers:
- robscott
title: Topology Aware Hints
feature:
  title: Topology Aware Hints
  description: >

>     Enables topology aware routing with topology hints on EndpointSlices.

Reviewer: This is too specific to be a front-page website feature, I'm afraid. Please _remove_ that feature card.

Author: Woops, didn't realize what this did. Was just copied from another page, thanks for catching this!

content_type: concept
weight: 10
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.21" state="alpha" >}}

_Topology Aware Hints_ enable topology aware routing with topology hints

> included in EndpointSlices. This approach tries to keep traffic close to where
> it originated from. This can result in decreased costs and improved performance.

Reviewer: nit; I'd write:
```suggestion
included in {{< glossary_tooltip text="EndpointSlices" term_id="endpoint-slice" >}}.
This approach tries to keep traffic closer to where it originated from. This can help to reduce costs
and improve performance.
```

Author: Updated, thanks!

<!-- body -->

> ## Introduction

Reviewer: nit; I'm afraid that following the overview with a section headed “introduction” is a bit of an anti-pattern.
How about “Motivation”?

Author: Good catch, also an artifact of a poorly copied page.

Kubernetes clusters are increasingly deployed in multi-zone environments.

> _Topology Aware Hints_ adds new functionality that can help keep traffic within

Reviewer: nit: in a concept page, we document the feature as if you're using it. That's different from a release note or a
release blog article where we highlight that the feature is new.

How about:
```suggestion
 _Topology Aware Hints_ provide a mechanism to help keep traffic within
```

Author: Updated, thanks!

the zone it originated from. This concept is commonly referred to as "Topology
Aware Routing". The "Hints" here refer to a key implementation detail. The

> EndpointSlice controller adds hints to EndpointSlices to allocate each endpoint

Reviewer: Are the hints labels? Annotations? Something else? Let's mention that here.

Author: Updated, but rather messy to get into the same sentence, feedback/suggestions appreciated here. This specifically populates `endpointSlice.endpoints[i].hints.zones[0].name`.

Reviewer: One thing to point out then is that the update is to `.endpoints.hints` and not to `.status.*` - right?

Author: Updated to say "on each endpoint" instead of "for each endpoint", hopefully that's clearer.

to a zone. Those hints are then consumed by components like kube-proxy as they
configure how requests are routed.

## Using Topology Aware Hints

> This feature can be enabled for individual Services by setting the

Reviewer: Style nit:
```suggestion
You can enable Topology Aware Hints for a Service by setting the
```

Author: Updated, thanks!

`service.kubernetes.io/topology-aware-hints` annotation to `auto`. This tells
the EndpointSlice controller to set topology hints if it is deemed safe.
Importantly, this does not guarantee that hints will always be set.

> ## How it Works

Reviewer: Optional improvement: I think it'd really help to have diagram showing the difference in traffic (or similar) for a Service with or without `service.kubernetes.io/topology-aware-hints`. After reading this page once I don't think I could draw that diagram but if it was already there I think my understanding would be a lot more thorough.

BTW, we can use Mermaid for diagrams so you don't need to use graphics software.

Author: Agree, this is a good idea, will try to set aside some time for this. Do you have any examples of diagrams that have been done well in k8s docs? Is Mermaid optional?

Reviewer: https://kubernetes.io/docs/concepts/services-networking/ingress/ is my usual example of diagrams in Mermaid.

https://kubernetes.io/docs/concepts/overview/components/ shows our “house style”; it uses icons from https://github.com/kubernetes/community/tree/master/icons

Reviewer: Actually I notice there's no icon for EndpointSlice yet.

Reviewer: It's OK to add the diagram later, eg when this is scheduled to go beta.

Author: Been overloaded this week so I don't think I'll get time to add this in for this release, but will keep it in mind for beta, agree that it would be helpful.

The functionality enabling this feature is split into two components: The
EndpointSlice controller and Kube-Proxy. This provides a high level overview of

> how each component implements this feature.

Reviewer: nit: consider omitting this or moving it much later in the page. So long as the feature works how we describe it, most readers won't need to look up which component implements which bit.

(also, you might use a network plugin that bypasses kube-proxy entirely but still honors these hints).

Author: Makes sense, I can work on this.

Author: I'm not sure how to actually make this more generic. I think it's important to understand that there are 2 components in play here, and both have various safeguards that could result in this feature not doing anything even when it appears to be enabled. I can understand why we might want to use "Control Plane" instead of "EndpointSlice Controller", but I don't think we have a sufficiently generic and broadly understood term for the component kube-proxy represents.

Reviewer: Ok, fair enough.

### EndpointSlice controller

The EndpointSlice controller is responsible for setting hints on EndpointSlices
when this feature is enabled. The controller allocates a proportional amount of
endpoints to each zone. This proportion is based on the number of CPU cores that
have been allocated to nodes running in that zone. For example, if one zone had
2 CPU cores and another zone only had 1 CPU core, the controller would allocated

> twice as many endpoints to the zone with 2 CPU cores.

Reviewer: I would actually title this “Resource matching” or otherwise word it to emphasize that the hints are based on a specific resource, `cpu`.

Does this count:
- the amount of vCPU you get from summing all the nodes in that zone?
- the amount of vCPU you get from summing all the nodes in that zone, minus system reservations?
- the amount of vCPU you get from summing the `cpu` requests of Pods in that zone matching the selector for the relevant Service?
-  the amount of vCPU you get from summing the `cpu` limits of Pods in that zone matching the selector for the relevant Service?
- something else?

(I couldn't work it out).

Author: This counts the amount of allocatable vCPU you get from summing all the nodes in that zone. So I think this is closest to your second bullet point, I'll work on cleaning this up.

Reviewer: https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/ might be relevant then?

Reviewer: (that diagram could use work &hellip;)

Author: Updated, thanks for the great reference!

### Kube-Proxy

Kube-Proxy filters the endpoints it routes to based on the hints set by the
EndpointSlice controller. Essentially, the controller allocates endpoints to
specific zones, and kube-proxy is responsible for ensuring that it only routes

> to endpoints with hints for the zone it is in.

Reviewer: Is it fair to say that the kube-proxy prefers same-zone endpoints and will try to send traffic to an endpoint in the same zone if the Service has added auto topology hints to the relevant EndpointSlice?

Reviewer: Could we see an example EndpointSlice with a hint added (eg as an excerpt of a YAML dump)?

Author: Yep, that's a good way to phrase this. I can rework this to add an example.

Author: Added an EndpointSlice example and reworked this description. There's still enough nuance that it's difficult to explain well. Open to any ideas to improve the wording here.

Reviewer: I'm happy that we don't need to iterate further on this part of the wording.

## Safeguards

Both the EndpointSlice controller and Kube-Proxy implementations have a number
of safeguards built in that will result in falling back to the original
approach. This results in traffic being routed to all endpoints, instead of just

> a topology aware subset. Here's what could trigger that:

Reviewer: nit: rather than “original” approach, let's come up with another name. What term would we pick if the feature had been graduated for a few releases?

Author: I've updated this to read "original cluster-wide" approach. I think it's going to be quite some time until topology becomes the default, so some variation of "original" is likely still valuable. Very open to other names/ideas though.

Reviewer: How about:

> The Kubernetes control plane and the kube-proxy on each node apply some safeguard rules before using Topology Aware Hints. If these don't check out, the kube-proxy selects endpoints from anywhere in your cluster, regardless of the zone.

Author: Updated, thanks!

1. **Insufficient number of endpoints:** If there are less endpoints than zones
   in a cluster, the controller will not assign any hints. 

2. **Impossible to achieve balanced allocation:** In some cases, it will be
   impossible to achieve a balanced allocation of endpoints among zones. For
   example, if zone-a is twice as large as zone-b, but there are only 2
   endpoints, an endpoint allocated to zone-a may receive twice as much traffic
   as zone-b. The controller wil not assign hints if it can't get this "expected
   overload" value below 20% for each zone. Importantly this is not based on
   real-time feedback. It is still possible for individual endpoints to become
   overloaded.

> 3. **One or more Nodes has insufficient information:** If a node does not have a

Reviewer: nit:
```suggestion
3. **One or more Nodes has insufficient information:** If any node does not have a
4. ```

Author: Updated, thanks!

>    `topology.kubernetes.io/zone` label or is not reporting a value for
>    allocatable CPU, this functionality will be disabled for all Services in a
>    cluster.

Reviewer: nit: using the active voice, we can explain _what_ does not act:
```suggestion
   allocatable CPU, the control plane does not set any topology-aware endpoint hints.
```

Author: Updated this, but not sure if it's clear enough yet. I'm not sure everyone will be able to connect "does not set any topology-aware hints" to "this feature will not do anything at all".

Reviewer: could add
> and so the kube-proxy selects endpoints from anywhere in your cluster, regardless of the zone.

Author: Updated, thanks!

4. **One or more endpoints does not have a zone hint:** When this happens,
   kube-proxy assumes that a transition from or to Topology Aware Hints is
   underway. Filtering endpoints for a Service in this state would be dangerous
   so Kube-Proxy falls back to using all endpoints.

5. **A zone is not represented in hints:** If kube-proxy is unable to find at

>    least one endpoing with a hint targeting the zone it is running in, it will

Reviewer 2: ```suggestion
   least one endpoint with a hint targeting the zone it is running in, it will
```

Author: Updated, thanks!

   fall back to using endpoints from all zones. This is most likely to happen as
   a new zone is being added to a cluster.

## Constraints

> * Topology Aware Hints are not compatible with `internalTrafficPolicy=Local`,

Reviewer 3: Is it invalid to spec `eTP: Cluster` and `iTP: Local` ?  I don't think that's disallowed?  In that case, topology hints modulate the meaning of "Cluster", no?

Author: You're completely right. There's a [bug in the implementation](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/topology.go#L30) that makes these exclusive, but that should not be the case.

Author: Follow up: https://github.com/kubernetes/kubernetes/issues/100313

Reviewer 4: if my understanding is correct, before the fix for https://github.com/kubernetes/kubernetes/issues/100313, it's not compatible with `externalTrafficPolicy=Local` as well?

Author: Updated, thanks!

  and therefore a Service cannot use both of these features. It is possible to
  use both features in the same cluster on different Services, just not on the
  same Service.

> * This approach will not work well for Services that have a large proportion of

Reviewer 3: The APPROACH might, but the implementation will need to adapt to handle it, right?

Reviewer: For an alpha feature, it's OK to document current shortcomings without making a commitment about future improvements to address them (we don't want to make promises unless we're sure of delivering on them).

Author: This is worth further discussion, but the only ways I can think to adapt to this would be new metrics that could enable feedback-based balancing or extra configuration.

Reviewer 3: What I meant is that the implementation could evolve to consider feedback and the `hints` API would still be appropriate

  traffic originating from a subset of zones. Instead this assumes that incoming
  traffic will be roughly proportional to the capacity of the Nodes in each
  zone.

* The EndpointSlice controller ignores unready nodes as it calculates the
  proportions of each zone. This could have unintended consequences if a large
  portion of nodes are unready.

> * The EndpointSlice controller does not take into account tolerations when

Reviewer: nit; how about:
```suggestion
* The EndpointSlice controller does not take into account {{< glossary_tooltip text="tolerations" term_id="toleration" >}} when
```

Author: Updated, thanks!

  deploying calculating the proportions of each zone. If the Pods backing a
  Service are limited to a subset of Nodes in the cluster, this will not be
  taken into account.

> * This may not work well with autoscaling. For example, if a lot of traffic is

Reviewer: Is this correct:
```suggestion
* Topology Aware hints may not work well if you use use horizontal autoscaling at either the pod or cluster level. For example, if a lot of traffic is
```
?

Author: I think this is primarily concerned with horizontal autoscaling at the pod level, but I think this could also interact poorly with vertical pod autoscaling, but I don't have as much experience there.

Reviewer: Feel free to pick any wording - emphasis on HPA makes sense given what you've said here.

Author: Good call, updated to add a reference to HPA here.

---
reviewers:
- robscott
title: Topology Aware Hints
content_type: concept
weight: 10
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.21" state="alpha" >}}

> _Topology Aware Hints_ enable topology aware routing with topology hints

Reviewer 3: This is self-referential. "X enables X".

Maybe "_Topology Aware Hints_ enable topology aware routing by including suggestions for how clients should consume endpoints" or something?

Author: Updated, thanks!

> included in {{< glossary_tooltip text="EndpointSlices" term_id="endpoint-slice"

Reviewer 3: will this linewrap work?  Or do you need to force it after the closing braces?

Author: Yep it did work, but I ended up removing this line as part of an earlier edit.

>}}. This approach tries to keep traffic closer to where it originated from.
This can help to reduce costs and improve performance.

<!-- body -->

## Motivation

Kubernetes clusters are increasingly deployed in multi-zone environments.
_Topology Aware Hints_ provides a mechanism to help keep traffic within the zone
it originated from. This concept is commonly referred to as "Topology Aware

> Routing". The "Hints" here refer to a key implementation detail. The

Reviewer 3: I'd strike the `The "Hints" here` sentence and segue with something like:  "When calculating the endpoints for a Service, the EndpointSlice controller considers the topology (region and zone) of each endpoint and populates the `hints` field to allocate it to a zone.  These hints ..."

Author: Updated, thanks!

EndpointSlice controller populates the hints field on each endpoint in an
EndpointSlice to allocate it to a zone. Those hints are then consumed by
components like kube-proxy as they configure how requests are routed.

## Using Topology Aware Hints

You can enable Topology Aware Hints for a Service by setting the

> `service.kubernetes.io/topology-aware-hints` annotation to `auto`. This tells

Reviewer 3: ignored

> `service.kubernetes.io/topology-aware-hints` annotation to `auto`. This tells

Reviewer 3: "Auto" or "auto"?  It should be "Auto" probably

Author: Unfortunately it is "auto", should I try to fix that in upstream? https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/utils.go#L395

Reviewer 3: Probably, and now you get to handle both!

Author: Covered by https://github.com/kubernetes/kubernetes/pull/100728

> `service.kubernetes.io/topology-aware-hints` annotation to `auto`. This tells

Author: Yep, the annotation is a way for users to opt into this feature on a Service, so dropping that opt-in would not be good here.

        - name: "zone-a"
```

### Kube-Proxy

Kube-Proxy filters the endpoints it routes to based on the hints set by the
EndpointSlice controller. In most cases, this means that kube-proxy will route
to endpoints in the same zone. Sometimes the controller allocates endpoints from
a different zone to ensure more even distribution of endpoints between zones.
This would result in some traffic being routed to other zones.

## Safeguards

The Kubernetes control plane and the kube-proxy on each node apply some
safeguard rules before using Topology Aware Hints. If these don't check out,
kube-proxy selects endpoints from anywhere in your cluster, regardless of the
zone.

1. **Insufficient number of endpoints:** If there are less endpoints than zones
   in a cluster, the controller will not assign any hints. 

2. **Impossible to achieve balanced allocation:** In some cases, it will be
   impossible to achieve a balanced allocation of endpoints among zones. For
   example, if zone-a is twice as large as zone-b, but there are only 2
   endpoints, an endpoint allocated to zone-a may receive twice as much traffic
   as zone-b. The controller wil not assign hints if it can't get this "expected

>    overload" value below 20% for each zone. Importantly this is not based on

Reviewer 3: I would not put the exact %age here - it's just going to get stale as the heuristic evolves :)

Author: Updated, thanks!

   real-time feedback. It is still possible for individual endpoints to become
   overloaded.

> 3. **One or more Nodes has insufficient information:** If any node does not have

Reviewer 5: Is this true even when a single node may be missing topology labels for a short period of time?

Reviewer 5: In other words, during that short period of time, would kube-proxy flap between topolgoy-aware hints and all endpoints?

Reviewer 3: Yes, that is what is is for now.

Author: Yep, the current approach is very quick to bail out. Would like to try to find some kind of acceptable threshold for missing data, but that's been hard to quantify so far.

---
reviewers:
- robscott
title: Topology Aware Hints
content_type: concept
weight: 10
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.21" state="alpha" >}}

_Topology Aware Hints_ enable topology aware routing by including suggestions

> for how clients should consume endpoints. This approach tries to keep traffic

Reviewer 6: Should we say more concretely?

"This approach ..."

=>

"This approach adds metadata to enable consumers of Endpoint(Slice) to be able to route traffic closer to where it is originated. For example, users can route traffic within a locality to reduce costs and improve performance."

Author: Updated, thanks!
